src/program.cpp

```diff
@@ -113,342 +86,115 @@ program& program::operator=(program p)
+std::unordered_map<std::string, shape> program::get_parameter_shapes() const
 {
-    assert(std::all_of(
-               args.begin(), args.end(), [&](instruction_ref x) { return has_instruction(x); }) &&
-           "Argument is not an exisiting instruction");
-    impl->instructions.push_back({builtin::returns{}, {}, args});
-    auto result = std::prev(impl->instructions.end());
-    instruction::backreference(result);
-    assert(result->valid(begin()));
-
-    return result;
+    const auto* mm = this->get_main_module();
+    return mm->get_parameter_shapes();
 }
 
-shape program::get_parameter_shape(std::string name) const
-{
-    auto ins = std::find_if(
-        impl->instructions.begin(), impl->instructions.end(), [&](const instruction& x) {
-            if(x.name() == "@param")
-            {
-                return any_cast<builtin::param>(x.get_operator()).parameter == name;
-            }
-            else
-            {
-                return false;
-            }
-        });
-    if(ins != this->end())
-        return ins->get_shape();
-    else
-        return {};
-}
-
-std::vector<std::string> program::get_parameter_names() const
+bool program::has_instruction(instruction_ref ins) const
 {
-    std::vector<std::string> result = impl->input_names;
-    std::unordered_set<std::string> params;
-    for(auto&& ins : impl->instructions)
-    {
-        if(ins.name() == "@param")
-        {
-            auto&& name = any_cast<builtin::param>(ins.get_operator()).parameter;
-            params.insert(name);
-        }
-    }
-    erase_if(result, [&](auto&& name) { return params.count(name) == 0; });
-    return result;
+    const auto* mm = this->get_main_module();
+    return mm->has_instruction(ins);
 }
 
-instruction_ref program::get_parameter(std::string name) const
-{
-    auto ins = std::find_if(
-        impl->instructions.begin(), impl->instructions.end(), [&](const instruction& x) {
-            if(x.name() == "@param")
-            {
-                return any_cast<builtin::param>(x.get_operator()).parameter == name;
-            }
-            else
-            {
-                return false;
-            }
-        });
-    if(ins != this->end())
-        return ins;
-    else
-        return this->end();
-}
+std::size_t program::size() const { return impl->modules.size(); }
 
-std::unordered_map<std::string, shape> program::get_parameter_shapes() const
+instruction_ref program::begin() const
 {
-    std::unordered_map<std::string, shape> result;
-    for(auto&& ins : impl->instructions)
-    {
-        if(ins.name() == "@param")
-        {
-            auto&& name  = any_cast<builtin::param>(ins.get_operator()).parameter;
-            result[name] = ins.get_shape();
-        }
-    }
-    return result;
+    const auto* mm = this->get_main_module();
+    return mm->begin();
 }
 
-bool program::has_instruction(instruction_ref ins) const
+instruction_ref program::end() const
```

Comment:
Program shouldn't have `begin` and `end` functions that return `instruction_ref`.

Reply:
I think a program actually contains a begin and end instruction, which is the begin() and end() of the main module. Do you mean we always need to use that of the main module?

Reply:
The program is a collection of modules(so begin and end should iterate over modules, but I dont think we need that) and the module is a collection of instructions.

> Do you mean we always need to use that of the main module?

Yes we should use begin and end from the module.

Reply:
fixed